src/main/java/frc/robot/subsystems/SwerveModule.java

```diff
@@ -103,15 +104,80 @@ public SwerveModule(
         driveMotorPID.setIntegratorRange(-C_MAX_VOLTAGE, C_MAX_VOLTAGE);
     }
 
+    private static void saveEncoderOffset() {
+
+        try {
+            BufferedWriter writer = new BufferedWriter(new FileWriter(C_ENCODER_OFFSETS_FILE_PATH));
+            for (int i = 0; i < 4; i++) {
+                writer.write(Double.toString(turnEncoderOffsets[i]));
+                writer.newLine();
+            }
+            writer.close();
+        } catch (IOException e) {
+            System.out.println("\u001b[31;1mFailed to write turn encoder offsets to file.\u001b[0m");
+        }
+    }
+
+    /**
+     * Gets the persisted encoder offset from the previous robot session.
+     *
+     * @return Absolute encoder's offset (in degrees) from 0 (forward).
+     */
+    private double getEncoderOffset() {
+        if (turnEncoderOffsets == null) {
+            turnEncoderOffsets = new double[4];
+            try {
+                BufferedReader reader = new BufferedReader(new java.io.FileReader(C_ENCODER_OFFSETS_FILE_PATH));
+                for (int i = 0; i < 4; i++) {
+                    turnEncoderOffsets[i] = Double.parseDouble(reader.readLine());
+                }
+                reader.close();
+            } catch (IOException e) {
+                System.out.println(
+                        "\u001b[31;1mFailed to read turn encoder offsets from file, please align wheels manually, "
+                                + "then reset encoders.\u001b[0m");
+
+                for (int i = 0; i < 4; i++) {
+                    turnEncoderOffsets[i] = 0;
+                }
+            }
+        }
+
+        return turnEncoderOffsets[moduleNum];
+    }
+
+    /**
+     * Sets the persisted encoder offset.
+     */
+    public void setEncoderOffset() {
+        if (turnEncoderOffsets == null) {
+            getEncoderOffset();
+        }
+
+        double currentAngle = m_turningMotor.getSelectedSensorPosition()
+                / Conversions.degreesToFalcon(1, Constants.SwerveModuleConstants.C_TURNING_MOTOR_GEAR_RATIO);
+        double offset = e_Encoder.getAbsolutePosition() - currentAngle;
+        turnEncoderOffsets[moduleNum] = offset;
+
+        // TODO: Make this not write to a file every time.
+        saveEncoderOffset();
```

Comment:
Is this TODO still relevant? Is there a time we want to save offsets without writing the file?

Reply:
I slightly misworded the TODO; that function runs 4 times lol, so that TODO is about not having it write to the file for every SwerveModule, not that it shouldn't run every shutdown.